Fill.py

```diff
@@ -129,21 +140,42 @@ def distribute_items_restrictive(world: MultiWorld, fill_locations=None):
         else:
             restitempool.append(item)
 
-    world.random.shuffle(fill_locations)
-    call_all(world, "fill_hook", progitempool, nonexcludeditempool, localrestitempool, nonlocalrestitempool, restitempool, fill_locations)
+    call_all(world, "fill_hook", progitempool, nonexcludeditempool,
+             localrestitempool, nonlocalrestitempool, restitempool, fill_locations)
+
+    locationDeficit = len(progitempool) - len(prioritylocations)
+    if locationDeficit > 0:
+        if locationDeficit > len(defaultlocations):
+            raise FillError(
+                f'Not enough locations for advancement items. There are {len(progitempool)} advancement items with {len(prioritylocations)} priority locations and {len(defaultlocations)} default locations')
+        prioritylocations += defaultlocations[:locationDeficit]
+        defaultlocations = defaultlocations[locationDeficit:]
+
+    fill_restrictive(world, world.state, prioritylocations, progitempool)
+    if prioritylocations:
+        defaultlocations = prioritylocations + defaultlocations
 
-    fill_restrictive(world, world.state, fill_locations, progitempool)
+    if progitempool:
+        fill_restrictive(world, world.state, defaultlocations, progitempool)
 
     if nonexcludeditempool:
-        world.random.shuffle(fill_locations)
-        fill_restrictive(world, world.state, fill_locations, nonexcludeditempool)  # needs logical fill to not conflict with local items
+        world.random.shuffle(defaultlocations)
+        # needs logical fill to not conflict with local items
+        nonexcludeditempool, defaultlocations = fast_fill(
```

Comment:
fill_restrictive has a side effect of marking locations as event locations. Changing this to use fast_fill instead of fill_restrictive will make it so that excluded locations are no longer considered event locations. I'm currently working to understand what effect this might have on progression balancing, but are there other consequences to this change?

Reply:
event flag effectively means this is a location which's item is already filled and logically relevant. If they aren't logically relevant then the flag should be safe to be omitted.